Display grade information

mtyaka opened the pull request:
## Description

Student's grade was not displayed on DnDv2 problems. This patch adds code to display score information beneath the display name of the problem (similar to CAPA).

It also implements the missing `max_score` method which is required by the LMS to correctly calculate and display grades on the "Progress" tab.

![screen shot 2016-10-20 at 10 10 22](https://cloud.githubusercontent.com/assets/32585/19551847/8150b9ca-96ad-11e6-86f9-612651fc91d9.png)
## JIRA
- [SOL-2030](https://openedx.atlassian.net/browse/SOL-2030)
- [SOL-2094](https://openedx.atlassian.net/browse/SOL-2094)
## Sandbox URLs

Latest commit on the sandboxe: 3ad5a184dfe6ecb6644603cc75973e19738a97da
- [DnD unit in LMS](http://pr13797.sandbox.opencraft.hosting/courses/course-v1:DnDX+DnD101+2016/courseware/ce77c57034524ffb938c2b16e8316b7d/d49d04ffc18f4e738229aab063234adf/1?activate_block_id=block-v1%3ADnDX%2BDnD101%2B2016%2Btype%40vertical%2Bblock%40631785c086ab45278755c34fa7562a31)
- [Progress page in LMS](http://pr13797.sandbox.opencraft.hosting/courses/course-v1:DnDX+DnD101+2016/progress)
- [DnD unit in Studio](http://studio-pr13797.sandbox.opencraft.hosting/container/block-v1:DnDX+DnD101+2016+type@vertical+block@631785c086ab45278755c34fa7562a31)
## Testing instructions
1. Add a new DnDv2 block using default settings (standard mode).
2. Open the problem in LMS. The grey line below the title should say: "1 point possible (graded)" or "1 point possible (ungraded)", depending on whether you created the problem in a graded or ungraded section.
3. Toggle section between graded and ungraded (in Course Outline view in Studio, click the settings icon of the subsection containing the DnDv2 problem and select "Not Graded" or any of the other options if you want the problem to be graded).
4. Refresh the page in LMS. Make sure the text changes between "(graded)" <-> "(ungraded)".
5. Open the "Progress" page. Verify that the section corresponding to the DnD problem says `0/1`.
6. Go back to the DnD problem and start dragging items to the board. As you drop each item, check that the grade updates accordingly. Also check that the grade on the "Progress" page is always in sync with the grade displayed on the DnD problem.
7. Reset the problem. Verify that resetting the problem does not reset the grade (both on the problem itself and on the "Progress" page).
8. Change max score (weight) setting in Studio. Delete student state to clear the existing grade, then check that the grades are scaled correctly for max_score.
9. Repeat the above steps with a DnD problem in "assessment" mode. In assessment mode, the grade should only update when you explicitly submit the problem, not on every drop like it does in standard mode.
## Notes and Concerns
#96 makes decoy items contribute to the final score. This has the surprising effect that the first move you do on a fresh DnD problem that includes at least one decoy item is wrong, you immediately get awarded points for the decoy item.
## Reviewers
- [x] OpenCraft: @itsjeyd
- [x] TNL: @cahrens and/or @staubina
- [ ] a11y: @edx/edx-accessibility
- [x] Product: @sstack22

edx-webhook commented:
Thanks for the pull request, @mtyaka! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=108&repo=edx-solutions%2Fxblock-drag-and-drop-v2

mtyaka commented on tests/unit/test_advanced.py:
```diff
@@ -123,51 +132,69 @@ def test_drop_item_correct(self):
 
         self.assertEqual(1, len(published_grades))
-        self.assertEqual({'value': 0.75, 'max_value': 1}, published_grades[-1])
+        self.assertEqual(0.75 * weight, self.block.grade)
```
Done.

mtyaka commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -444,8 +442,51 @@ function DragAndDropTemplates(configuration) {
+                    ctx.max_score
+                );
+            }
+        }
```
You're right, that function was getting way too long. I moved the progress template to a separate function.

mtyaka commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -463,6 +504,7 @@ function DragAndDropTemplates(configuration) {
         return (
             h('section.themed-xblock.xblock--drag-and-drop', [
                 problemTitle,
+                h('div.problem-progress', {key: progress_id, id: progress_id}, progress_text),
```
@itsjeyd The `key` entry does not translate to HTML, it's an internal property used by virtual-dom so that it can distinguish elements when reordering them.
For example if you first render two elements `[h('div', 'a'), h('div', 'b')]`, and then reorder them when rendering the second time (`[h('div', 'b'), h('div', 'a')]`), virtual-dom will not be able to know that the elements have been reordered so it will patch the two corresponding DOM elements rather than reordering them. It usually doesn't matter since the visual end result is the same, but if the two DOM elements contain any internal state (for example if you modify then with jQuery outside of virtual-dom), it can lead to weird issues.

There is no reordering going on here, so the `key` property isn't really needed. I added it while trying to fix weird selenium failures just in case it helped (it didn't; the issues were not caused by virtual-dom). I will remove the `key` property.

itsjeyd replied:
@mtyaka Sounds good, thanks!

mtyaka commented:
@itsjeyd 

> One question about a11y: Using Orca and Firefox I was able to get the screen reader to announce the contents of the problem-progress element to me (as part of reading the problem-title). But updates to the grade were not announced to me as I was placing items on the board in standard mode, or after clicking "Submit" in assessment mode. Do CAPA problems do this? If so, we should probably try and replicate that behavior for DnDv2.1.

Good idea, thank you. CAPA problems display the grade in the feedback area when you submit the problem and the screen reader reads that.

DnDv2 doesn't display the grade in the feedback area, and in standard mode the grade updates dynamically while you're working on the problem, so it does make sense for the screen reader to announce updates to the score. I added `role="status"` and `aria-live="polite"` to the progress element and verified that VoiceOver announced the changes as I work on a DnDv2 problem.

itsjeyd commented on tests/integration/test_interaction.py:
```diff
@@ -271,6 +272,29 @@ def test_final_feedback_and_reset(self, action_key):
+        total_items = len(items_with_zones) + len(items_without_zones)
+
+        progress = self._page.find_element_by_css_selector('.problem-progress')
+        self.assertEqual(progress.text, '1 point possible (ungraded)')
```
@mtyaka Thanks for the explanation, I'm OK with leaving the tests as is.

itsjeyd commented:
Thanks for the updates, @mtyaka! I had another look at the code and have no more comments. Let me know when the sandbox is up to date so I can do another quick round of testing.

mtyaka commented:
@itsjeyd The sandbox should be up to date :)

itsjeyd commented:
@mtyaka I gave this another quick test on the sandbox, and it's working fine. But my screen reader is acting up right now, so I couldn't verify that changes to the grade are getting announced when adding items to the board in standard mode or when submitting a solution in assessment mode.

@haikuginger Since you're sprint firefighter, could you help out and verify just that aspect on the sandbox please? See https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/108#issuecomment-255314435 for reference. I'll approve this PR as soon as that's done.

mtyaka commented:
@haikuginger In assessment mode only negative feedback popups are shown. Positive feedback is still displayed in the feedback message area, but there is no popup. I think it's the popup that somehow prevents the update to the score from being announced.

@itsjeyd I added code to display grade information in the feedback area when in assessment mode. In assessment mode we already displayed the highest grade when user runs out of attempts, but after this change, the grade is displayed after every submission.

The sandbox has been updated, please take a look.

mtyaka commented:
@cahrens or @staubina, @edx/edx-accessibility: This is ready for your review.

cahrens commented:
@mtyaka I created https://openedx.atlassian.net/browse/TNL-5822 for the review of this PR. I put it into our current sprint, but I can't guarantee super-fast turnaround on this one, as I wasn't aware that any more DnD PR reviews were coming.

Should we expect more reviews in the next 2 weeks?

mtyaka commented:
@cahrens Thank you for creating the ticket! I don't think we plan to do any more work on DnD in the next two weeks, but I'll ask @bradenmacdonald to confirm.

staubina commented:
@mtyaka 

The link labeled "DnD unit in LMS" is no longer working.  I will follow the steps above to test these changes out.

staubina commented:
I was able to get to the problem by updating the dates the course was run.  I am guessing this was changed for testing.

staubina commented:
Is it expected that I get 0.2 points when I answer the problem incorrectly on my first attempt?

<img width="612" alt="screen shot 2016-10-25 at 11 39 25 am" src="https://cloud.githubusercontent.com/assets/2854941/19692924/bf1cff04-9aa7-11e6-9034-49d0763ab905.png">

staubina commented:
If I perform the following steps I would expect my score to drop.  Currently it stays at its highest and never goes down.

1. Drop an answer in the correct location
2. Submit (review score)
3. Move that answer to an incorrect location
4. Submit (review score)

Expect: 
Score to drop when I move my answer to the incorrect location
Actual: 
Score stays the same.

What is the expected behavior here from the AC?

staubina commented on tests/integration/test_interaction_assessment.py:
```diff
@@ -220,15 +221,22 @@ def test_do_attempt_feedback_is_updated(self):
         expected_feedback = "\n".join(feedback_lines)
         self.assertEqual(self._get_feedback().text, expected_feedback)
 
+        # Place the item into incorrect zone. The score does not change.
```
Is this comment true?  I am seeing different behavior as I mentioned in a previous comment.

bradenmacdonald commented:
@cahrens 
> Should we expect more reviews in the next 2 weeks?

I'm still getting back to speed on everything but AFAIK, no.

Long answer: The only remaining items are not going to come up anytime soon:
* Comprehensive a11y review + follow up (SOL-1978): this hasn't been scheduled yet, so isn't part of our current sprint.
* https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/104 which is a very low priority fix and not part of any upcoming sprints (SOL-2088)
* any handover coordination requested to support TNL with maintenance of this block (if that's still the plan) (SOL-2095)

nasthagiri commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -172,6 +172,12 @@ class DragAndDropBlock(XBlock, XBlockWithSettingsMixin, ThemableXBlockMixin):
+        """
+        Return the problem's max score. Required by the grading system in the LMS.
+        """
+        return self.weight
```
Please note that `max_score` is different from `weight`.  The weight is already applied when calculating the problem's score.  So there's no need to specify it in the max_score value as well.  In fact, this may result in unexpected value.  See the formula at https://openedx.atlassian.net/wiki/display/TNL/Robust+Grades+Design#RobustGradesDesign-ProblemScores.

nasthagiri replied:
In your case, max_score may always be `1`.

mtyaka replied:
Thanks you @nasthagiri, I missed that! I will change `max_score` to always return `1`.

sstack22 commented:
@mtyaka - this looks great! Just one note - when a problem weight = 0, we also note it as ungraded (even if it is within a graded subsection).

mtyaka commented:
Thanks @sstack22! I will update the code so that weight = 0 displays as "ungraded" as well.

mtyaka commented:
@staubina Thanks a lot for the quick review!

> Is it expected that I get 0.2 points when I answer the problem incorrectly on my first attempt?

The reason is that the demo problem includes one decoy item which counts towards the final score (see "Notes and Concerns" section in PR description). Even if you first move is incorrect, the decoy item is in the correct position (the item bank), which is why you get awarded 0.2 points. The behavior is surprising, but I can't think of an easy ways to work around that.

> If I perform the following steps I would expect my score to drop. Currently it stays at its highest and never goes down. What is the expected behavior here from the AC?

The expected behavior is to display the highest achieved score (see https://openedx.atlassian.net/browse/SOL-2030?focusedCommentId=211959&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-211959). This is different from CAPA problems.

mtyaka commented:
I changed `max_score` to always return `1`,  removed grade-related feedback in assessment mode when `weight == 0`, and rebuilt the sandbox with the latest code.

staubina commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -443,8 +441,59 @@ function DragAndDropTemplates(configuration) {
+        if (ctx.grade !== null) {
+            if (is_graded) {
+                progress_template = ngettext(
+                    '{earned}/{possible} point (graded)',
```
Add Translation comments here, as shown here: http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/internationalization/i18n.html#python-source-code

mtyaka replied:
@staubina Translations added, thanks for the suggestion.

staubina commented:
@sstack22 @mtyaka We made a recent update to Capa to show a score of zero with a weight of 1 as 0/1 point.  This does not seem to be the case here.  Should we update this PR to match that case?

mtyaka commented:
@staubina 

> We made a recent update to Capa to show a score of zero with a weight of 1 as 0/1 point. This does not seem to be the case here. Should we update this PR to match that case?

I think it makes sense to display the zero score as 0/weight. The latest commit does that.
I also updated the sandbox with the latest code.

staubina commented:
@mtyaka Thank you for making these changes.  This PR Looks good to me. :+1:

mtyaka commented:
Thanks @staubina!

@sstack22 Do you want to take another look?

staubina commented:
@mtyaka, @nasthagiri Has more experience with the grader and would to take a look.  

@nasthagiri this is ready for your review.

nasthagiri commented on tests/integration/test_interaction.py:
```diff
@@ -271,6 +272,29 @@ def test_final_feedback_and_reset(self, action_key):
     def test_keyboard_help(self, use_keyboard):
         self.interact_with_keyboard_help(use_keyboard=use_keyboard)
 
+    def test_grade_display(self):
```
Is there also an acceptance test that makes sure the grade for DND problems are now displayed on the LMS' progress page?

haikuginger replied:
@nasthagiri - taking over for @mtyaka on this temporarily.

We test for the grade event in tests/integration/test_events.py.

nasthagiri replied:
sounds good.  thanks.

nasthagiri commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -443,8 +441,63 @@ function DragAndDropTemplates(configuration) {
+                );
+            }
+            progress_template = progress_template.replace('{earned}', formatNumber(ctx.grade));
+        } else {
```
Why is this `else` clause needed?  It doesn't look like there are any valid cases for it.

haikuginger replied:
This clause renders a different message in cases where the user hasn't yet submitted an answer.

nasthagiri commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -172,6 +172,13 @@ class DragAndDropBlock(XBlock, XBlockWithSettingsMixin, ThemableXBlockMixin):
 
+    def max_score(self):  # pylint: disable=no-self-use
+        """
+        Return the problem's max score, which for DnDv2 always equals 1.
```
Note: Currently, you are defining the `max_score` as 1.  However, in the future, you may consider having the `max_score` be equal to the number of items that need to be dragged.  This way, when DND is used as an assessment, learners can get partial credit for some of the items they got correct.

haikuginger replied:
@nasthagiri, let me look into this one. I know we emit a different `max_value` [in our grade events](https://github.com/open-craft/xblock-drag-and-drop-v2/blob/3ad5a184dfe6ecb6644603cc75973e19738a97da/drag_and_drop_v2/drag_and_drop_v2.py#L697), so I'm not sure why we're returning a static value for the `max_score` method. There may be some legacy considerations here.

e-kolpakov replied:
@nasthagiri I believe the same have already [been discussed](https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/108#discussion_r85107013). DnDv2 calculates student grade as a [ratio of correct item placements to total count of items](https://github.com/open-craft/xblock-drag-and-drop-v2/blob/3ad5a184dfe6ecb6644603cc75973e19738a97da/drag_and_drop_v2/drag_and_drop_v2.py#L898), so max_score is one, but `grade = student_score * weight`.

We might consider making max_score equal to total number of items, and adjust grade calculations accordingly, but I'm concerned that this might be not backward compatible.

I'm not sure about this part:

> learners can get partial credit for some of the items they got correct.

If I'm not mistaking learners already receive partial credit if they haven't achieved 100% score.

Also, maybe there's a typo in [the formula](https://openedx.atlassian.net/wiki/display/TNL/Robust+Grades+Design#RobustGradesDesign-ProblemScores), but `weight` is in nominator and denominator, so it should cancel out, and the formula should become `learner_raw_score / max_score`.

nasthagiri replied:
The formula intentionally puts the `weight` in both the numerator and the denominator.  The reason is because the `weighted_possible` (denominator) needs to be the total possible points, including the `weight`.  And therefore, the `weighted_earned` (numerator) is also multiplied by the `weight` in order to keep the total score the same.

nasthagiri commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -666,9 +682,10 @@ def _mark_complete_and_publish_grade(self):
         # ... and from higher grade to lower
-        if grade > self.grade:
+        current_grade = self._get_grade_if_set()
+        if current_grade is None or grade > current_grade:
```
Hmmm.. Does this mean if the learner changes their answer so that it's different from the correct answer, they'll see their new incorrect answer, but their grade remains the same?  Won't that be confusing to the learner?  It's also inconsistent with CAPA problems.  FYI @sstack22.
(The sandbox is not responding currently so I can't verify this.)

e-kolpakov replied:
@nasthagiri @sstack22 keeping the highest score is by design:
* [if you have implemented already, no worries at all"](https://openedx.atlassian.net/browse/SOL-1944?focusedCommentId=205655&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-205655))
* and than [latest version implemented just keeps track of highest achieved score and only submits if the new score was higher](https://openedx.atlassian.net/browse/SOL-1944?focusedCommentId=205666&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-205666)

Than the discussion around this stopped. We might want to revisit it, but I would suggest changing that in a separate PR if necessary.

sstack22 replied:
Just to follow up on this - from a product perspective, I'm OK with keeping the max score as it was designed. I think this benefits learners. We haven't heard feedback from partners/learners thus far, but we can certainly reconsider in the future. Agreed that if this changes in the future, it would be a separate story.

nasthagiri replied:
I LOVE the new DragNDrop.  In my testing of it on prod, I do find it confusing that once my "completion" is locked in - and then I choose to continue to experiment/test with the block, I stop getting feedback on my changes.  Any then it becomes _impossible_ for me to validate which of my changes is the correct answer.  Have you considered locking the UI once the learner "completes" the block in the "Assessment" mode - as is currently done in the "Standard" mode of the block?

e-kolpakov replied:
@nasthagiri I believe it was discussed at least once. I can't find any links to prove that, but as far as I can recall not locking UI after correct attempt was intentional - to allow student to experiment with the problem more without the risk of getting lower grade (i.e. to read error item feedback).

nasthagiri commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -443,8 +441,63 @@ function DragAndDropTemplates(configuration) {
+        // Formats a number to 4 decimals without trailing zeros
+        // (1.00 -> '1'; 1.50 -> '1.5'; 1.333333333 -> '1.3333').
+        var formatNumber = function(n) { return n.toFixed(4).replace(/\.?0+$/, ''); };
+        var is_graded = ctx.graded && ctx.max_score > 0;
```
Does `ctx.max_score` include the `weight` of the problem?  It should - to allow course teams to mark problems as not part of the grade.

nasthagiri replied:
Suggestion: Consider having more descriptive variable names in this code so the reader can easily distinguish between `weighted_max_score` versus `raw_max_score`, for example.
(We did a similar renaming exercise recently in the grading infrastructure in the platform for readability and to prevent bugs.)

e-kolpakov commented:
@nasthagiri I've addressed/answered remaining questions - could you please take another look?

sstack22 commented:
👍  on my end, given the issue I noted was resolved.

nasthagiri commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -443,8 +441,63 @@ function DragAndDropTemplates(configuration) {
+                );
+            }
+            progress_template = progress_template.replace('{earned}', formatNumber(ctx.grade));
+        } else {
```
I think this comment was missed.  If we do not verify that `ctx.weighted_max_score > 0` here, then we'll still be displaying scoring information when the course team sets the problem's weight to 0.

FYI @sstack22.  From my perspective, showing "0 points possible" can be confusing.  However, that's currently what CAPA does.  So I'm not flagging this as a merge-blocker.  However, we should consider whether this makes sense for all problem types.

e-kolpakov replied:
@sstack22 I believe this is a product decision, so please decide and create a task in JIRA if modifications are needed.

nasthagiri commented on tests/unit/test_basics.py:
```diff
@@ -86,13 +89,25 @@ def test_get_configuration(self):
+        Test that the methods required by the LMS grading interface work as expected.
+        """
+        self.block.weight = weight
+        # Max score is different from weight and should always equals 1 for drag and drop problems.
```
Nit: typo - "equals" -> "equal"

e-kolpakov commented:
@mtyaka I've merged this PR - please delete the branch if you don't need it anymore.

nasthagiri commented:
Thanks @e-kolpakov.

mtyaka commented:
@e-kolpakov Thanks, branch deleted.

e-kolpakov commented:
@mtyaka One more thing - looks like we both forgot to bump version number, so I've submitted another PR #109 to do that - if you could approve it it would be great :)
